### PR TITLE
unicode support

### PR DESCRIPTION
**Pull request by FinchPowers**

### CODE REVIEW

File: edlib/src/edlib.cpp [hunk 1]
```diff
@@ -1269,6 +1272,42 @@ static int obtainAlignmentHirschberg(
...
+
+    int res = static_cast<int>(c);
+    for (; remaining > 0; --remaining, ++idx) {
+        c = in[idx];
+        if ((c & 0xC0) != 0x80) {
+            throw std::runtime_error(
+                "Invalid UTF-8 character sequence encountered in buffer at index "
+                + to_string(idx));
+        }
+        res += static_cast<int>(c);
```
Comment (simlmx) [inline edlib/src/edlib.cpp]: C'est aussi simple que de les additionner? faut pas genre tasser d'un byte et copier les 8 bits du caractère?

File: edlib/src/edlib.cpp [hunk 2]
```diff
@@ -1269,6 +1272,42 @@ static int obtainAlignmentHirschberg(
     return EDLIB_STATUS_OK;
 }
 
+static int getCodePoint(const char * in, int & idx) {
```
Comment (simlmx) [inline edlib/src/edlib.cpp]: Peut-être l'appeler getCodePointAndIncrementIndex?

File: edlib/src/edlib.cpp [hunk 3]
```diff
@@ -1287,42 +1326,61 @@ static int obtainAlignmentHirschberg(
...
-    *queryTransformed = (unsigned char *) malloc(sizeof(unsigned char) * queryLength);
-    *targetTransformed = (unsigned char *) malloc(sizeof(unsigned char) * targetLength);
+    *queryTransformed = (int *) malloc(sizeof(int) * queryLength);
+    *targetTransformed = (int *) malloc(sizeof(int) * targetLength);
 
     // Alphabet information, it is constructed on fly while transforming sequences.
-    unsigned char letterIdx[128]; //!< letterIdx[c] is index of letter c in alphabet
-    bool inAlphabet[128]; // inAlphabet[c] is true if c is in alphabet
-    for (int i = 0; i < 128; i++) inAlphabet[i] = false;
+    constexpr int ALPHABET_ARR_SIZE = 128;
```
Comment (simlmx) [inline edlib/src/edlib.cpp]: peut-être l'appeler ASCII_ALPHABET_SIZE?

**Review by simlmx — COMMENTED** [review]
sinon si c'est bien testé dans MLDB +1

**Comment by jeremybarnes** [comment]
This doesn't make sense to me.

If the input code points are integers (which should really be char32_t, by the way) as the change in signatures would indicate, then the input should be pure Unicode code points (pre-decoded) and we wouldn't need any kind of Unicode decoding.

If the input code points are characters, then the Unicode decoding is needed, but then the signatures of the functions shouldn't change.

As it is, you pass them as 32 bit integers as if they were Unicode code points, but you then decode them as if they weren't.

**Comment by FinchPowers** [comment]
int\* -> char32_t*: done

We receive unicode encoded in a char array. We extract the code points out of it and we use that representation for all the operations. Does that make sense?

File: edlib/src/edlib.cpp [hunk 4]
```diff
@@ -54,51 +55,51 @@ struct Block {
...
+static inline Word* buildPeq(int alphabetLength, const char32_t* query, int queryLength);
 
 
 
 /**
  * Main edlib method.
  */
-EdlibAlignResult edlibAlign(const char* queryOriginal, const int queryLength,
-                            const char* targetOriginal, const int targetLength,
+EdlibAlignResult edlibAlign(const char* queryOriginal, int queryLength,
```
Comment (FinchPowers) [inline edlib/src/edlib.cpp:101]: This is the entry point, the function called by MLDB.

File: edlib/src/edlib.cpp [hunk 5]
```diff
@@ -1269,6 +1272,42 @@ static int obtainAlignmentHirschberg(
     return EDLIB_STATUS_OK;
 }
 
+static char32_t getCodePoint(const char * in, int & idx) {
```
Comment (FinchPowers) [inline edlib/src/edlib.cpp]: Clearly broken and badly named atm.

**Comment by jeremybarnes** [comment]
A couple of other thoughts:
- One of the customers that uses edlib also has performance problems.  This change may make that worse, as we quadruple the memory required.  Since MLDB knows whether or not a string has Unicode or not, we may want to benchmark and if it's slower consider including two versions of the function (ASCII 8 bit and Unicode 32 bit)
- Are we planning on upstreaming these changes?  If so, we should talk to the original developer and get their input on the problem and how we've solved it

**Comment by FinchPowers** [comment]
- On top of that "throw" call have been added to a C styled library so it **will** leak whenever a bad char is encountered. I should fix that too.

**Comment by FinchPowers** [comment]
- Benchmarking through MLDB shows no difference of performance.
- I replaced the throw with char32_t 0xFFFFFFFF, an unexisting unicode char.

**Comment by FinchPowers** [comment]
- I don't plan to upstream my changes because the official way to compile it is pre c++11. Thus tuples and char32_t don't exist.

**Comment by FinchPowers** [comment]
@jeremybarnes Review please.

**Comment by jeremybarnes** [comment]
It's OK for the moment.  +1.  If customers still have performance problems we may want to use SSE4.2 string instructions, which need 8 bit characters, and so we may want to move to a template interface.  But until we know that it's fine to merge.

**Pull request merged**